sky/ssh_node_pools/state.py

```diff
@@ -0,0 +1,134 @@
+"""The database for ssh node pool information."""
+import functools
+import os
+from pathlib import Path
+import threading
+from typing import List, Optional
+
+import sqlalchemy
+from sqlalchemy import exc as sqlalchemy_exc
+from sqlalchemy import orm
+
+from sky import sky_logging
+from sky.ssh_node_pools import constants
+from sky.ssh_node_pools import models
+from sky.utils import common_utils
+from sky.utils import db_utils
+
+# TODO(kyuds): we need to support Postgres in the future.
+# currently, only using sqlite3
+
+logger = sky_logging.init_logger(__name__)
+
+_SQLALCHEMY_ENGINE: Optional[sqlalchemy.engine.Engine] = None
+_DB_INIT_LOCK = threading.Lock()
+
+
+def _create_table():
+    # refer to `global_user_state.py` create_table()
+    if _SQLALCHEMY_ENGINE is None:
+        raise ValueError('ssh node pool sqlalchemy engine is None')
+
+    if (_SQLALCHEMY_ENGINE.dialect.name
+            == db_utils.SQLAlchemyDialect.SQLITE.value and
+            not common_utils.is_wsl()):
+        try:
+            with orm.Session(_SQLALCHEMY_ENGINE) as session:
+                session.execute(sqlalchemy.text('PRAGMA journal_mode=WAL'))
+                session.commit()
+        except sqlalchemy_exc.OperationalError as e:
+            if 'database is locked' not in str(e):
+                raise
+
+    # create table if they don't exist
+    models.SSHBase.metadata.create_all(bind=_SQLALCHEMY_ENGINE)
+
+
+def _initialize_and_get_db() -> sqlalchemy.engine.Engine:
+    global _SQLALCHEMY_ENGINE
+    if _SQLALCHEMY_ENGINE is not None:
+        return _SQLALCHEMY_ENGINE
+
+    with _DB_INIT_LOCK:
+        if _SQLALCHEMY_ENGINE is None:
+            db_path = os.path.expanduser(constants.SKYSSH_DB_PATH)
+            Path(db_path).parents[0].mkdir(parents=True, exist_ok=True)
+            _SQLALCHEMY_ENGINE = sqlalchemy.create_engine('sqlite:///' +
+                                                          db_path)
```

Comment:
```suggestion
            conn_string = None
            if os.environ.get(constants.ENV_VAR_IS_SKYPILOT_SERVER) is not None:
                conn_string = skypilot_config.get_nested(('db',), None)
            if conn_string:
                logger.debug(f'using db URI from {conn_string}')
                engine = sqlalchemy.create_engine(conn_string,
                                                  poolclass=sqlalchemy.NullPool)
            else:
                pathlib.Path(db_path).parents[0].mkdir(parents=True,
                                                       exist_ok=True)
                engine = sqlalchemy.create_engine('sqlite:///' + db_path)
```
This should add support for pSQL

Reply:
done